storage/range.go

```diff
@@ -649,26 +721,132 @@ func (r *Range) processRaftCommand(idKey cmdIDKey, index uint64,
 		// This command originated elsewhere so we must create a new reply buffer.
 		reply = args.CreateReply()
 	}
-	err := r.executeCmd(index, args, reply)
+
+	err := r.applyRaftCommand(index, multiraft.NodeID(raftCmd.OriginNodeID), args, reply)
+
 	if cmd != nil {
 		cmd.done <- err
 	} else if err != nil {
-		log.Errorf("error executing raft command %s: %s", method, err)
+		log.Errorf("error executing raft command %s: %s", args.Method(), err)
 	}
 	return err
 }
 
-// startGossip periodically gossips the cluster ID if it's the
-// first range and the raft leader.
-func (r *Range) startGossip() {
-	r.stopper.RunWorker(func() {
+// applyRaftCommand applies a raft command from the replicated log to
+// the underlying state machine (i.e. the engine).
+func (r *Range) applyRaftCommand(index uint64, originNodeID multiraft.NodeID, args proto.Request, reply proto.Response) error {
+	if index <= 0 {
+		log.Fatalf("raft command index is <= 0")
+	}
+	header := args.Header()
+
+	// Check the response cache to ensure idempotency.
+	if proto.IsWrite(args) {
+		if ok, err := r.respCache.GetResponse(header.CmdID, reply); ok || err != nil {
+			if ok { // this is a replay! extract error for return
+				log.V(1).Infof("found response cache entry for %+v", args.Header().CmdID)
+				return reply.Header().GoError()
+			}
+			// In this case there was an error reading from the response
+			// cache. Instead of failing the request just because we can't
+			// decode the reply in the response cache, we proceed as though
+			// idempotence has expired.
+			log.Errorf("unable to read result for %+v from the response cache: %s", args, err)
```

Comment:
Well it would mean corruption in the stored value. Doesn't seem a panic-able offense as the process itself isn't corrupted by this event. Perhaps we should return an error directly to the client. I'm undecided.